python/ml4ir/data/tfrecord_reader.py

```diff
@@ -68,76 +65,117 @@ def _parse_sequence_example_fn(sequence_example_proto):
         features_dict = dict()
 
         # Explode context features into all records
-        for feat, t in context.items():
-            t = tf.expand_dims(t, axis=0)
-            t = tf.tile(t, multiples=[max_num_records])
+        for feature_info in feature_config.get_context_features():
+            feature_name = feature_info["name"]
+            feature_layer_info = feature_info.get("feature_layer_info")
+
+            feature_tensor = context_features.get(feature_name)
+
+            feature_tensor = tf.expand_dims(feature_tensor, axis=0)
+            feature_tensor = tf.tile(feature_tensor, multiples=[max_num_records])
 
             # If feature is a string, then decode into numbers
-            if feature_config.get_dict()[feat]["type"] == FeatureTypeKey.STRING:
-                t = io.decode_raw(
-                    t,
+            if feature_layer_info["type"] == FeatureTypeKey.STRING:
+                feature_tensor = io.decode_raw(
+                    feature_tensor,
                     out_type=tf.uint8,
-                    fixed_length=feature_config.get_dict()[feat]["max_length"],
+                    fixed_length=feature_layer_info["max_length"],
```

Comment:
do we enforce anywhere that `feature_layer_info["max_length"]` is set?

Reply:
No. We should. Should probably go in `FeatureConfig`. We should have a validate step. I'm going to add a `validate()` method in the `FeatureConfig` class, so that we fail fast. We will populate it with more validation as we go.